Conceptual_Guide/Part_8-semantic_caching/README.md

```diff
@@ -0,0 +1,328 @@
+    has an integration in Faiss, more on this can be found [here.](https://docs.rapids.ai/api/cuvs/nightly/integrations/faiss/)
+* [Theine](https://github.com/Yiling-J/theine): High performance in-memory
+cache.
+    - We will use it as our exact match cache backend. After the most similar
+    prompt is identified, the corresponding cached response id extracted from
+    the cache. This library supports multiple eviction policies, in this
+    tutorial we use "LRU".
+    - One may also look into [MemCached](https://memcached.org/about) as a
+    potential alternative.
+
+Provided [script](./artifacts/semantic_caching.py) is heavily annotated and we
+encourage users to look through the code to gain better clarity in all
+the necessary stages.
+
+## Incorporating Semantic Cache into your workflow
+
+For this tutorial, we'll use the [vllm backend](https://github.com/triton-inference-server/vllm_backend)
+as our example, focusing on demonstrating how to cache responses for the
+non-streaming case. The principles covered here can be extended to handle
+streaming scenarios as well.
+
+### Cutomising vllm backend
+
+First, let's start by cloning Triton's vllm backend repository. This will
+provide the necessary codebase to implement our semantic caching example.
+
+``bash
+git clone https://github.com/triton-inference-server/vllm_backend.git
+```
+
+With the repository cloned, the next step is to add the
+[semantic_caching.py.](./artifacts/semantic_caching.py) script to
+the appropriate directory. This script contains the logic for our semantic
+caching implementation.
+
+```bash
+wget -P vllm_backend/src/utils/ https://raw.githubusercontent.com/triton-inference-server/tutorials/refs/heads/main/Conceptual_Guide/Part_8-semantic_caching/artifacts/semantic_caching.py
+```
+
+Now that we have added the semantic caching script, let's proceed by making
+some adjustments in `/vllm_backend/src/model.py`. These changes will integrate
+the semantic caching functionality into the model.
+
+First, ensure that you import the necessary classes from `semantic_caching.py`:
+
+```diff
+...
+
+from utils.metrics import VllmStatLogger
++from utils.semantic_caching import SemanticCPUCacheConfig, SemanticCPUCache
+```
+
+Next, initialize the semantic cache during the initialization step.
+This setup will prepare your model to utilize semantic caching during
+its operations.
+
+```diff
+    def initialize(self, args):
+        self.args = args
+        self.logger = pb_utils.Logger
+        self.model_config = json.loads(args["model_config"])
+        ...
+
+        # Starting asyncio event loop to process the received requests asynchronously.
+        self._loop = asyncio.get_event_loop()
+        self._event_thread = threading.Thread(
+            target=self.engine_loop, args=(self._loop,)
+        )
+        self._shutdown_event = asyncio.Event()
+        self._event_thread.start()
++       self.semantic_cache = SemanticCPUCache(SemanticCPUCacheConfig)
```

Comment:
This should probably be one of the following (correct if wrong):

Add `()` after `SemanticCPUCacheConfig` to instantiate config object
```diff
+       self.semantic_cache = SemanticCPUCache(SemanticCPUCacheConfig()) 
```

or instantiate config separately for clarity:
```diff
+       config = SemanticCPUCacheConfig(...)
+       self.semantic_cache = SemanticCPUCache(config)
```

Reply:
Good point, suggestions are what was intended to be done. The reason why it worked during testing, is I believe because I define all default values for the `dataclass`, it just accesses these default class attributes. e.g. :
```
>>> 
>>> @dataclass
... class SemanticCPUCacheConfig:
...     """
...     Configuration class for SemanticCPUCache.
... 
...     Attributes:
...         cache (Any): The cache object to use.
...         encoder (Any): The encoder object for embedding queries.
...         index (Any): The index object for similarity search.
...         threshold (float): The similarity threshold for considering a match.
...         key_mapper (Any): The key mapper object for managing key-ID mappings.
...     """
...     cache: Any = "Cache"
...     encoder: Any = "SentenceTransformer"
...     index: Any = "faiss.IndexIDMap(faiss.IndexFlatL2(384))"
...     threshold: float = 0.25
...     key_mapper: Any = 123
... 
>>> 
>>> SemanticCPUCacheConfig
<class '__main__.SemanticCPUCacheConfig'>
>>> c = SemanticCPUCacheConfig
>>> c
<class '__main__.SemanticCPUCacheConfig'>
>>> c.threshold
0.25
>>> c.index
'faiss.IndexIDMap(faiss.IndexFlatL2(384))'
>>> c.key_mapper
123
>>> c.cache
'Cache'
```